Editorial pass (#129)
decentralized-identity/bbs-signature: draft-bbs-signatures.md

: returns the point P corresponding to the canonical representation ostr, or INVALID if ostr is not a valid output of point\_to\_octets.  This operation is also known as deserialization.

subgroup\_check(P) -> VALID or INVALID

> : returns VALID when the point P is an element of the subgroup of order p, and INVALID otherwise. This function can always be implemented by checking that p \* P is equal to the identity element.  In some cases, faster checks may also exist, e.g., [@Bowe19].

Author: ```suggestion
: returns VALID when the point P is an element of the subgroup of order p, and INVALID otherwise. This function can always be implemented by checking that q \* P is equal to the identity element.  In some cases, faster checks may also exist, e.g., [@Bowe19].
```

> : returns VALID when the point P is an element of the subgroup of order p, and INVALID otherwise. This function can always be implemented by checking that q \* P is equal to the identity element.  In some cases, faster checks may also exist, e.g., [@Bowe19].

Author: ```suggestion
: returns VALID when the point P is an element of the subgroup of order q, and INVALID otherwise. This function can always be implemented by checking that q \* P is equal to the identity element.  In some cases, faster checks may also exist, e.g., [@Bowe19].
```

For readability, this document makes these transformations implicitly, but they MUST precede every call to the HASH or XOF function.

> Optional input/parameters to operations that feature in a call to a HASH or XOF function, that are not supplied to the operation should be omitted from the call, for example if X is an optional input/parameter whilst A and B are required then the procedural step of `HASH(A || X || B)` MUST be evaluated to `HASH(A || B)`.

Reviewer: ```suggestion
Optional input/parameters to operations that feature in a call to a HASH or XOF function, that are not supplied to the operation should default to the empty string. For example, if X is an optional, not supplied, input/parameter whilst A and B are required, then the procedural step of `HASH(A || X || B)` MUST be evaluated to `HASH(A || "" || B)`.
```
Since, in some places where we hash optional values we specify that if not supplied they default to the empty string, the above might be more accurate??

Author: What about empty octet string to be more specific?

Reviewer: Yeap. That sounds better.

8. return SK
```

> **Note** This operation is the RECOMMENDED way of generating a secret key, but its use is not required for compatibility, and implementations MAY use a different KeyGen procedure. For security, such an alternative MUST output a secret key that is statistically close to uniformly random in the range 1 <= SK < r.

Reviewer: ```suggestion
**Note** This operation is the RECOMMENDED way of generating a secret key, but its use is not required for compatibility, and implementations MAY use a different KeyGen procedure. For security, such an alternative MUST output a secret key that is statistically close to uniformly random in the range 0 < SK < r.
```
We use the 0 < SK < r noation above i think.

### Sign

This operation computes a deterministic signature from a secret key (SK), over a header and a vector of messages.

```
signature = Sign(SK, header, (msg_1,..., msg_L), (H_1,..., H_L))

Inputs:

- SK (REQUIRED), an octet string of the form outputted by the KeyGen operation.
- header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.

> - msg_1,..., msg_L (OPTIONAL), a vector of octet strings. Application specific messages.

Reviewer: Not really a big fun of defining the messages as "Application specific". It's definitely correct, however, it may raise the question of "what other messages are there??"

Author: Have removed for now, we can further refine this language later.

- H_1,..., H_L (OPTIONAL), points of G1. Generators used to commit each message.

Parameters:

- Ciphersuite_ID (REQUIRED), ASCII string. The unique ID of the ciphersuite.
- H_s (REQUIRED), point of G1. The generator for the blinding value of the signature.
- H_d (REQUIRED), point of G1. The generator used to sign the signature domain.

Outputs:

- signature, an octet string.

Procedure:

> 1. PK = SkToPk(SK)

Reviewer: Although i like the simpler API with not having to supply the PK this may add unnecessary computations?? The PK will most likely be cached. It doesn't really guard against un trusted issuers either (they can just skip this step).

Author: Im happy to keep supplying the public key to the operation API, provided we add a consideration that implementations SHOULD validate the relationship between SK and PK, does that work as a compromise?

Reviewer: If we are worried about issuers making mistakes then yes a consideration sounds as a better option. Maybe another option could also be to make the relationship clear in the definition of the inputs in Sign? Something like;

```
- SK (REQUIRED), an octet string of the form outputted by the KeyGen operation.
- PK (REQUIRED), an octet string encoded public key output from SkToPk on input the above SK.
```

### ProofVerify

> This operation checks that a proof is valid for a header, vector of revealed messages (along side their corresponding originally index position when signed) and presentation header against a public key (PK).

Reviewer: ```suggestion
This operation checks that a proof is valid for a header, vector of revealed messages (along side their index corresponding to their original position when signed) and presentation header against a public key (PK).
```
Just a minor suggestion though the original is also completely fine. So really not necessary to use the alternative.
